pkg/cincinnati/cincinnati.go

```diff
@@ -146,17 +156,98 @@ func (c Client) GetUpdates(ctx context.Context, uri *url.URL, arch string, chann
+			return current, nil, nil, &Error{
+				Reason:  "ResponseInvalid",
+				Message: fmt.Sprintf("invalid recommended update node: %s", err),
+			}
+		}
+		updates = append(updates, update)
+	}
+
+	var conditionalUpdates []configv1.ConditionalUpdate
+	for _, conditionalEdges := range graph.ConditionalEdges {
+		for _, edge := range conditionalEdges.Edges {
+			if version.String() == edge.From {
+				var target *node
+				for i, node := range graph.Nodes {
+					if node.Version.String() == edge.To {
+						target = &graph.Nodes[i]
+						break
+					}
+				}
+				if target == nil {
+					return current, updates, nil, &Error{
+						Reason:  "ResponseInvalid",
+						Message: fmt.Sprintf("no node for conditional update %s", edge.To),
+					}
+				}
+				update, err := convertRetrievedUpdateToRelease(*target)
+				if err != nil {
+					return current, updates, nil, &Error{
+						Reason:  "ResponseInvalid",
+						Message: fmt.Sprintf("invalid conditional update node: %s", err),
+					}
+				}
+				conditionalUpdates = append(conditionalUpdates, configv1.ConditionalUpdate{
+					Release: update,
+					Risks:   conditionalEdges.Risks,
+				})
+			}
+		}
+	}
+
+	for i := len(updates) - 1; i >= 0; i-- {
+		for _, conditionalUpdate := range conditionalUpdates {
+			if conditionalUpdate.Release.Image == updates[i].Image {
+				klog.Warningf("Update to %s listed as both a conditional and unconditional update; preferring the conditional update.", conditionalUpdate.Release.Version)
+				updates = append(updates[:i], updates[i+1:]...)
+			}
+		}
+	}
+
+	if len(updates) == 0 {
+		updates = nil
+	}
+
+	for i := len(conditionalUpdates) - 1; i >= 0; i-- {
+		for j, risk := range conditionalUpdates[i].Risks {
+			conditionalUpdates[i].Risks[j].MatchingRules, err = clusterconditions.PruneInvalid(ctx, risk.MatchingRules)
+			if len(conditionalUpdates[i].Risks[j].MatchingRules) == 0 {
+				klog.Warningf("Conditional update to %s, risk %q, has empty pruned matchingRules; dropping this target to avoid upsetting the Kubernetes API server.  Pruning results: %s", conditionalUpdates[i].Release.Version, risk.Name, err)
+				conditionalUpdates = append(conditionalUpdates[:i], conditionalUpdates[i+1:]...)
+			} else if err != nil {
+				klog.Warningf("Conditional update to %s, risk %q, has pruned matchingRules (although other valid, recognized matchingRules were given, and are sufficient to keep the conditional update): %s", conditionalUpdates[i].Release.Version, risk.Name, err)
+			}
+		}
+	}
+
+	targets := make(map[string]int, len(conditionalUpdates))
+	for _, conditionalUpdate := range conditionalUpdates {
+		targets[conditionalUpdate.Release.Image]++
 	}
 
-	return current, updates, nil
+	for i := len(conditionalUpdates) - 1; i >= 0; i-- {
+		if targets[conditionalUpdates[i].Release.Image] > 1 {
+			klog.Warningf("Upstream declares %d conditional updates to %s; dropping them all.", targets[conditionalUpdates[i].Release.Image], conditionalUpdates[i].Release.Version)
```

Comment:
> Should this be an error with a more drastic consequence such that we would expect to find and correct during CI rather than ignoring and fielding?

I don't mind checking for this error in graph-data presubmits, to help us keep our own data clean.  But my understanding based on [the discriminating union discussion][1] was that @deads2k preferred "silently hide confusing data" over "bubble up 'hey, this graph-data I got from upstream isn't all that great' where the cluster admin might notice out" (which was my preference and original proposal).  See also [here][2]:

> David expects graph-admins to be keeping watch over enough ClusterVersion content that they notice quickly if the CVO is not publishing some data that they expected to get through. So that in cases where partial data does slip into graph-data, the graph-data admins are likely to notice and fix the data before [cluster] admins have time to get too worked up.

So while I'm still in favor of complaining somewhere the admin is more likely to see, I don't feel empowered to make that sort of change without getting David's sign-off on the scope and mechanics.

[1]: https://github.com/openshift/enhancements/blob/2cc2d9b331532c852878a7c793f3a754914c824e/enhancements/update/targeted-update-edge-blocking.md#discriminating-union-implementation
[2]: https://github.com/openshift/enhancements/blob/2cc2d9b331532c852878a7c793f3a754914c824e/enhancements/update/targeted-update-edge-blocking.md#metadata-to-include-when-blocking-a-conditional-request